File: code-recipes/java/src/main/java/easyship/EasyShipOrderProcessingRecipe.java
package easyship;

import software.amazon.spapi.api.externalfulfillment.shipments.v2024_09_11.ShipmentProcessingApi;
import software.amazon.spapi.api.externalfulfillment.shipments.v2024_09_11.ShipmentRetrievalApi;
import software.amazon.spapi.models.externalfulfillment.shipments.v2024_09_11.*;
import util.Recipe;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.List;

/**
 * Easy Ship Order Processing Recipe
 * ==================================
 * Workflow: Fetch → Acknowledge → Package → Shipping Options → Labels
 * 
 * This recipe demonstrates the complete workflow for processing Amazon Easy
 * Ship orders
 * using the External Fulfillment API (also known as SmartConnect API).
 */
public class EasyShipOrderProcessingRecipe extends Recipe {

    private final ShipmentRetrievalApi shipmentRetrievalApi;
    private final ShipmentProcessingApi shipmentProcessingApi;
    private final String marketplaceId;
    private final PackageStatus shipmentStatus;

    private Shipment shipment;
    private String packageId;
    private String shippingOptionId;

    public EasyShipOrderProcessingRecipe() {
        this("A1PA6795UKMFR9", PackageStatus.CREATED); // DE marketplace
    }

    public EasyShipOrderProcessingRecipe(String marketplaceId, PackageStatus shipmentStatus) {
        this.marketplaceId = marketplaceId;
        this.shipmentStatus = shipmentStatus;
        this.packageId = "PKG-" + System.currentTimeMillis();

        this.shipmentRetrievalApi = new ShipmentRetrievalApi.Builder()
                .lwaAuthorizationCredentials(lwaCredentials)
                .endpoint(util.Constants.BACKEND_URL)
                .build();

        this.shipmentProcessingApi = new ShipmentProcessingApi.Builder()
                .lwaAuthorizationCredentials(lwaCredentials)
                .endpoint(util.Constants.BACKEND_URL)
                .build();
    }

    /**
     * Check if shipment is Easy Ship (shippingType=MARKETPLACE, channelName=MFN).
     */
    private boolean isEasyShipShipment(Shipment shipment) {
        try {
            return "MARKETPLACE".equals(shipment.getShippingInfo().getShippingType().getValue()) &&
                    "MFN".equals(shipment.getMarketplaceAttributes().getChannelName());
        } catch (Exception e) {
            return false;
        }
    }

    // -------------------------------------------------------------------------
    // Step 1: Fetch open shipments
    // -------------------------------------------------------------------------

    private ShipmentsResponse fetchOpenShipments() {
        System.out.println("\n--- Step 1: Fetching Easy Ship orders ---");

        try {
            ShipmentsResponse response = shipmentRetrievalApi.getShipments(
                    shipmentStatus.getValue(),
                    marketplaceId,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null);

            List<Shipment> easyShipShipments = new ArrayList<>();
            if (response.getShipments() != null) {
                for (Shipment s : response.getShipments()) {
                    if (isEasyShipShipment(s)) {
                        easyShipShipments.add(s);
                    }
                }
            }

            if (!easyShipShipments.isEmpty()) {
                this.shipment = easyShipShipments.get(0);
                System.out.println("✅ Found " + easyShipShipments.size() +
                        " Easy Ship order(s) - Shipment ID: " + this.shipment.getId());
            } else {
                System.out.println("⚠️  No Easy Ship orders found");
            }

            return response;
        } catch (Exception e) {
            System.err.println("Error fetching shipments: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    // -------------------------------------------------------------------------
    // Step 2: Acknowledge shipment
    // -------------------------------------------------------------------------

    private void acknowledgeShipment() {
        System.out.println("\n--- Step 2: Acknowledging shipment ---");

        if (this.shipment == null) {
            System.out.println("⚠️  Skipping: No shipment to confirm was found");
            return;
        }

        try {
            shipmentProcessingApi.processShipment(this.shipment.getId(), "CONFIRM", null);
            System.out.println("✅ Shipment " + this.shipment.getId() + " confirmed");
        } catch (Exception e) {
            System.err.println("Error acknowledging shipment: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    // -------------------------------------------------------------------------
    // Step 3: Create packages
    // -------------------------------------------------------------------------

    private void createPackages() {
        System.out.println("\n--- Step 3: Creating shipment packages ---");

        if (this.shipment == null) {
            System.out.println("⚠️  Skipping: No shipment available");
            return;
        }

        if (this.shipment.getLineItems() == null || this.shipment.getLineItems().isEmpty()) {
            System.out.println("⚠️  Skipping: No line items in shipment");
            return;
        }

        try {
            // Get recommended package dimensions and weight from shipment
            PackageDimensions packageDimensions;
            Weight packageWeight;

            if (this.shipment.getShippingInfo() != null &&
                    this.shipment.getShippingInfo().getRecommendedPackages() != null &&
                    !this.shipment.getShippingInfo().getRecommendedPackages().isEmpty()) {

                RecommendedPackage recommendedPackage = this.shipment.getShippingInfo()
                        .getRecommendedPackages().get(0);
                packageDimensions = recommendedPackage.getDimensions();
                packageWeight = recommendedPackage.getWeight();

                System.out.println("Using recommended dimensions: " +
                        packageDimensions.getLength().getValue() + "x" +
                        packageDimensions.getWidth().getValue() + "x" +
                        packageDimensions.getHeight().getValue() + " " +
                        packageDimensions.getLength().getDimensionUnit().getValue());
                System.out.println("Using recommended weight: " +
                        packageWeight.getValue() + " " + packageWeight.getWeightUnit().getValue());
            } else {
                System.out.println("⚠️  No recommended package dimensions found, using defaults");
                // Fallback to default dimensions
                Dimension length = new Dimension();
                length.setDimensionUnit(Dimension.DimensionUnitEnum.CM);
                length.setValue(Float.toString(28.0f));

                Dimension width = new Dimension();
                width.setDimensionUnit(Dimension.DimensionUnitEnum.CM);
                width.setValue(Float.toString(26.0f));

                Dimension height = new Dimension();
                height.setDimensionUnit(Dimension.DimensionUnitEnum.CM);
                height.setValue(Float.toString(10.0f));

                packageDimensions = new PackageDimensions();
                packageDimensions.setLength(length);
                packageDimensions.setWidth(width);
                packageDimensions.setHeight(height);

                packageWeight = new Weight();
                packageWeight.setValue(Float.toString(280.0f));
                packageWeight.setWeightUnit(Weight.WeightUnitEnum.G);
            }

            // Build package line items from shipment line items
            PackageLineItem packageLineItem = new PackageLineItem();
            packageLineItem.setPackageLineItemId(this.shipment.getLineItems().get(0).getShipmentLineItemId());
            packageLineItem.setQuantity(1);

            ModelPackage modelPackage = new ModelPackage();
            modelPackage.setId(this.packageId);
            modelPackage.setDimensions(packageDimensions);
            modelPackage.setWeight(packageWeight);

            PackageLineItems packageLineItems = new PackageLineItems();
            packageLineItems.add(packageLineItem);

            modelPackage.setPackageLineItems(packageLineItems);

            Packages packagesBody = new Packages();
            packagesBody.setPackages(List.of(modelPackage));

            shipmentProcessingApi.createPackages(packagesBody, this.shipment.getId());

            System.out.println("✅ Package created successfully");
            System.out.println("Package ID: " + this.packageId);
        } catch (Exception e) {
            System.err.println("Error creating packages: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    // -------------------------------------------------------------------------
    // Step 4: Retrieve available shipping options
    // -------------------------------------------------------------------------

    private ShippingOptionsResponse retrieveShippingOptions() {
        System.out.println("\n--- Step 4: Retrieving shipping options ---");

        if (this.shipment == null || this.packageId == null) {
            System.out.println("⚠️  Skipping: Missing shipment or package ID");
            return null;
        }

        try {
            ShippingOptionsResponse response = shipmentProcessingApi.retrieveShippingOptions(
                    this.shipment.getId(),
                    this.packageId);

            // For simplicity, this recipe always uses the recommended shipping option if
            // available
            if (response.getRecommendedShippingOption() != null) {
                this.shippingOptionId = response.getRecommendedShippingOption().getShippingOptionId();
                System.out.println("✅ Using recommended shipping option");
                System.out.println("Shipping Option ID: " + this.shippingOptionId);
                if (response.getRecommendedShippingOption().getCarrierName() != null) {
                    System.out.println("Carrier: " + response.getRecommendedShippingOption().getCarrierName());
                }
            } else if (response.getShippingOptions() != null && !response.getShippingOptions().isEmpty()) {
                // Fallback to first option if no recommendation
                ShippingOptions firstShippingOptions = response.getShippingOptions().get(0);
                this.shippingOptionId = firstShippingOptions.getShippingOptionId();

                System.out.println("✅ Found " + response.getShippingOptions().size() + " shipping option(s)");
                System.out.println("Selected Shipping Option ID: " + this.shippingOptionId);
            } else {
                System.out.println("⚠️  No shipping options available");
            }

            return response;
        } catch (Exception e) {
            System.err.println("Error retrieving shipping options: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    // -------------------------------------------------------------------------
    // Step 5: Generate shipping labels
    // -------------------------------------------------------------------------

    private ShipLabelsResponse generateShippingLabels() {
        System.out.println("\n--- Step 5: Generating shipping labels ---");

        if (this.shipment == null || this.packageId == null || this.shippingOptionId == null) {
            System.out.println("⚠️  Skipping: Missing required IDs");
            return null;
        }

        try {
            // Build request body using model class
            ShipLabelsInput labelsInput = new ShipLabelsInput();
            labelsInput.setPackageIds(List.of(this.packageId));

            ShipLabelsResponse response = shipmentProcessingApi.generateShipLabels(
                    this.shipment.getId(),
                    "GENERATE",
                    labelsInput,
                    this.shippingOptionId);

            if (response.getPackageShipLabelList() != null && !response.getPackageShipLabelList().isEmpty()) {
                System.out.println("✅ Shipping labels generated successfully");
                System.out.println("Number of labels: " + response.getPackageShipLabelList().size());

                // Get the first label for simplicity
                PackageShipLabel label = response.getPackageShipLabelList().get(0);
                System.out.println("\nPackage ID: " + label.getPackageId());
                System.out.println("Tracking ID: " + label.getShipLabelMetadata().getTrackingId());
                System.out.println("Carrier: " + label.getShipLabelMetadata().getCarrierName());

                // Download and open the shipping label
                downloadAndOpenLabel(label.getFileData());
            } else {
                System.out.println("⚠️  No labels generated");
            }

            return response;
        } catch (Exception e) {
            System.err.println("Error generating shipping labels: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    /**
     * Download and open the shipping label from the presigned URL.
     * 
     * The file type is determined from the Content-Type header in the response:
     * - application/zpl: ZPL format (for thermal printers)
     * - application/pdf: PDF format
     * - image/png: PNG format
     * - text/plain: Plain text format
     * 
     * In production, ZPL labels should be sent to a thermal printer.
     * For this recipe, we download and open the label for demonstration.
     */
    private void downloadAndOpenLabel(DocumentV2 fileData) {
        try {
            System.out.println("Downloading label...");

            // Download the label and get content type from response header
            URI uri = new URI(fileData.getUrl());
            HttpURLConnection connection = (HttpURLConnection) uri.toURL().openConnection();
            connection.setRequestMethod("GET");

            String contentType = connection.getContentType();
            if (contentType == null) {
                contentType = "application/octet-stream";
            }

            // Determine file extension from content type
            String fileExtension;
            String labelFormat;

            if (contentType.toLowerCase().contains("zpl")) {
                fileExtension = ".zpl";
                labelFormat = "ZPL";
            } else if (contentType.toLowerCase().contains("pdf")) {
                fileExtension = ".pdf";
                labelFormat = "PDF";
            } else if (contentType.toLowerCase().contains("png")) {
                fileExtension = ".png";
                labelFormat = "PNG";
            } else {
                fileExtension = ".txt";
                labelFormat = "TEXT";
            }

            System.out.println("Label Format: " + labelFormat + " (Content-Type: " + contentType + ")");

            // Save to temporary file
            Path tempFile = Files.createTempFile("shipping_label_", fileExtension);

            try (InputStream inputStream = connection.getInputStream()) {
                Files.copy(inputStream, tempFile, StandardCopyOption.REPLACE_EXISTING);
            }

            System.out.println("✅ Label saved to: " + tempFile.toAbsolutePath());

            // In production: Send ZPL files to thermal printer
            if ("ZPL".equals(labelFormat)) {
                System.out.println("ℹ️  Production: Send ZPL file to thermal printer");
            }

            // For this recipe: Open the downloaded file
            System.out.println("🖼️  Opening label...");
            if (java.awt.Desktop.isDesktopSupported()) {
                java.awt.Desktop.getDesktop().open(tempFile.toFile());
            }

        } catch (Exception e) {
            System.err.println("❌ Error: " + e.getMessage());
        }
    }

    // -------------------------------------------------------------------------
    // Main recipe entry point
    // -------------------------------------------------------------------------

    @Override
    protected void start() {
        System.out.println("======================================================================");
        System.out.println("Easy Ship Order Processing Recipe");
        System.out.println("======================================================================");

        // Step 1: Fetch open orders
        fetchOpenShipments();

        // Step 2: Acknowledge shipment
        acknowledgeShipment();

        // Step 3: Create packages
        createPackages();

        // Step 4: Retrieve shipping options
        retrieveShippingOptions();

        // Step 5: Generate shipping labels
        generateShippingLabels();

        System.out.println("\n======================================================================");
        System.out.println("✅ Easy Ship order processing completed successfully");
        System.out.println("======================================================================");
    }
}

File: code-recipes/python/src/recipes/easyship/easyship_order_processing_recipe.py
"""
Easy Ship Order Processing Recipe
==================================
Workflow: Fetch → Acknowledge → Package → Shipping Options → Labels
"""

from typing import Optional
import time
import urllib.request
import webbrowser
import tempfile
import os

from spapi import SPAPIClient, SPAPIConfig
from spapi.api.external_fulfillment_shipments_v2024_09_11 import shipment_processing_api
from spapi.api.external_fulfillment_shipments_v2024_09_11 import shipment_retrieval_api
from spapi.models.external_fulfillment_shipments_v2024_09_11.dimension import Dimension
from spapi.models.external_fulfillment_shipments_v2024_09_11.package import Package
from spapi.models.external_fulfillment_shipments_v2024_09_11.package_dimensions import PackageDimensions
from spapi.models.external_fulfillment_shipments_v2024_09_11.package_line_item import PackageLineItem
from spapi.models.external_fulfillment_shipments_v2024_09_11.packages import Packages
from spapi.models.external_fulfillment_shipments_v2024_09_11.ship_labels_input import ShipLabelsInput
from spapi.models.external_fulfillment_shipments_v2024_09_11.ship_labels_response import ShipLabelsResponse
from spapi.models.external_fulfillment_shipments_v2024_09_11.shipments_response import ShipmentsResponse
from spapi.models.external_fulfillment_shipments_v2024_09_11.shipping_options_response import ShippingOptionsResponse
from spapi.models.external_fulfillment_shipments_v2024_09_11.status import Status
from spapi.models.external_fulfillment_shipments_v2024_09_11.weight import Weight
from spapi.models.external_fulfillment_shipments_v2024_09_11.package_status import PackageStatus
from src.util.recipe import Recipe
from src import config


class EasyShipOrderProcessingRecipe(Recipe):
    """Process Easy Ship orders through the External Fulfillment API."""

    def __init__(
        self,
        config: Optional[SPAPIConfig] = None,
        shipment_retrieval_api: Optional[shipment_retrieval_api.ShipmentRetrievalApi] = None,
        shipment_processing_api: Optional[shipment_processing_api.ShipmentProcessingApi] = None,
        marketplace_id: str = "A1PA6795UKMFR9",  # DE marketplace
        shipment_status: str = PackageStatus.CREATED,
    ) -> None:
        super().__init__(config=config)
        self._shipment_retrieval_api = shipment_retrieval_api
        self._shipment_processing_api = shipment_processing_api
        self.marketplace_id = marketplace_id
        self.shipment_status = shipment_status
        
        self.shipment = None
        self.package_id = f"PKG-{int(time.time())}"
        self.shipping_option_id: Optional[str] = None

    @property
    def shipment_retrieval_api(self) -> shipment_retrieval_api.ShipmentRetrievalApi:
        if self._shipment_retrieval_api is None:
            client = SPAPIClient(
                self.config, 
                oauth_endpoint=f"{config.backend_url}/auth/o2/token",
                endpoint=config.backend_url
            )
            self._shipment_retrieval_api = shipment_retrieval_api.ShipmentRetrievalApi(client.api_client)
        return self._shipment_retrieval_api
    
    @property
    def shipment_processing_api(self) -> shipment_processing_api.ShipmentProcessingApi:
        if self._shipment_processing_api is None:
            client = SPAPIClient(
                self.config, 
                oauth_endpoint=f"{config.backend_url}/auth/o2/token",
                endpoint=config.backend_url
            )
            self._shipment_processing_api = shipment_processing_api.ShipmentProcessingApi(client.api_client)
        return self._shipment_processing_api

    def is_easy_ship_shipment(self, shipment) -> bool:
        """Check if shipment is Easy Ship (shippingType=MARKETPLACE, channelName=MFN)."""
        try:
            return (shipment.shipping_info.shipping_type == "MARKETPLACE" and 
                    shipment.marketplace_attributes.channel_name == "MFN")
        except AttributeError:
            return False

    # -------------------------------------------------------------------------
    # Step 1: Fetch open shipments
    # -------------------------------------------------------------------------
    
    def fetch_open_shipments(self) -> ShipmentsResponse:
        """Step 1: Fetch and filter Easy Ship orders."""
        print("\n--- Step 1: Fetching Easy Ship orders ---")
        
        response = self.shipment_retrieval_api.get_shipments(
            status=self.shipment_status,
            marketplace_id=self.marketplace_id,
        )
        
        easy_ship_shipments = [s for s in (response.shipments or []) if self.is_easy_ship_shipment(s)]
        
        if easy_ship_shipments:
            self.shipment = easy_ship_shipments[0]
            print(f"✅ Found {len(easy_ship_shipments)} Easy Ship order(s) - Shipment ID: {self.shipment.id}")
        else:
            print(f"⚠️  No Easy Ship orders found")
        
        return response

    # -------------------------------------------------------------------------
    # Step 2: Acknowledge shipment
    # -------------------------------------------------------------------------
    
    def acknowledge_shipment(self) -> None:
        """Step 2: Confirm shipment."""
        print("\n--- Step 2: Acknowledging shipment ---")
        
        if not self.shipment:
            print("⚠️  Skipping: No shipment to confirm was found")
            return
        
        self.shipment_processing_api.process_shipment(shipment_id=self.shipment.id, operation="CONFIRM")
        print(f"✅ Shipment {self.shipment.id} confirmed")
        
    # -------------------------------------------------------------------------
    # Step 3: Create packages
    # -------------------------------------------------------------------------

    def create_packages(self) -> None:
        """
        Provide package dimensions, weight, and line items for the shipment.
        Uses recommended package dimensions and weight from the shipment data.
        """
        print("\n--- Step 3: Creating shipment packages ---")
        
        if not self.shipment:
            print("⚠️  Skipping: No shipment available")
            return
        
        if not self.shipment.line_items or len(self.shipment.line_items) == 0:
            print("⚠️  Skipping: No line items in shipment")
            return
        
        # Get recommended package dimensions and weight from shipment
        try:
            recommended_package = self.shipment.shipping_info.recommended_packages[0]
            package_dimensions = recommended_package.dimensions
            package_weight = recommended_package.weight
            
            print(f"Using recommended dimensions: "
                  f"{package_dimensions.length.value}x{package_dimensions.width.value}x{package_dimensions.height.value} "
                  f"{package_dimensions.length.dimension_unit}")
            print(f"Using recommended weight: {package_weight.value} {package_weight.weight_unit}")
        except (AttributeError, IndexError):
            print("⚠️  No recommended package dimensions found, using defaults")
            # Fallback to some default dimensions
            package_dimensions = PackageDimensions(
                length=Dimension(dimension_unit="CM", value=28),
                width=Dimension(dimension_unit="CM", value=26),
                height=Dimension(dimension_unit="CM", value=10)
            )
            package_weight = Weight(value=280, weight_unit="grams")
        
        # Build package line items from shipment line items
        package_line_item = PackageLineItem(
            package_line_item_id=self.shipment.line_items[0].shipment_line_item_id,
            quantity=1
        )
        
        package = Package(
            id=self.package_id,
            dimensions=package_dimensions,
            weight=package_weight,
            package_line_items=[package_line_item]
        )
        
        packages_body = Packages(packages=[package])
        
        self.shipment_processing_api.create_packages(
            shipment_id=self.shipment.id,
            body=packages_body
        )
        
        print("✅ Package created successfully")
        print(f"Package ID: {self.package_id}")

    # -------------------------------------------------------------------------
    # Step 4: Retrieve available shipping options
    # -------------------------------------------------------------------------

    def retrieve_shipping_options(self) -> Optional[ShippingOptionsResponse]:
        """
        Get the list of shipping options provided by the marketplace.
        
        Returns:
            ShippingOptionsResponse containing available shipping options.
        """
        print("\n--- Step 4: Retrieving shipping options ---")
        
        if not self.shipment or not self.package_id:
            print("⚠️  Skipping: Missing shipment or package ID")
            return None
        
        response: ShippingOptionsResponse = self.shipment_processing_api.retrieve_shipping_options(
            shipment_id=self.shipment.id,
            package_id=self.package_id
        )
        
        # For simplicity, this recipe always uses the recommended shipping option if available
        if response.recommended_shipping_option:
            self.shipping_option_id = response.recommended_shipping_option.shipping_option_id
            print(f"✅ Using recommended shipping option")
            print(f"Shipping Option ID: {self.shipping_option_id}")
            if hasattr(response.recommended_shipping_option, 'carrier_name'):
                print(f"Carrier: {response.recommended_shipping_option.carrier_name}")
        elif response.shipping_options and len(response.shipping_options) > 0:
            # Fallback to first option if no recommendation
            first_option = response.shipping_options[0]
            self.shipping_option_id = first_option.id
            
            print(f"✅ Found {len(response.shipping_options)} shipping option(s)")
            print(f"Selected Shipping Option ID: {self.shipping_option_id}")
        else:
            print("⚠️  No shipping options available")
        
        return response

    # -------------------------------------------------------------------------
    # Step 5: Generate shipping labels
    # -------------------------------------------------------------------------

    def generate_shipping_labels(self) -> Optional[ShipLabelsResponse]:
        """
        Generate and retrieve shipping labels for the packages.
        
        Returns:
            ShipLabelsResponse containing generated shipping labels.
        """
        print("\n--- Step 5: Generating shipping labels ---")
        
        if not self.shipment or not self.package_id or not self.shipping_option_id:
            print("⚠️  Skipping: Missing required IDs")
            return None
        
        # Build request body using model class
        labels_input = ShipLabelsInput(package_ids=[self.package_id])
        
        response: ShipLabelsResponse = self.shipment_processing_api.generate_ship_labels(
            shipment_id=self.shipment.id,
            operation="GENERATE",
            shipping_option_id=self.shipping_option_id,
            body=labels_input
        )
        
        if response.package_ship_label_list:
            print("✅ Shipping labels generated successfully")
            print(f"Number of labels: {len(response.package_ship_label_list)}")
            
            # Get the first label for simplicity
            label = response.package_ship_label_list[0]
            print(f"\nPackage ID: {label.package_id}")
            print(f"Tracking ID: {label.ship_label_metadata.tracking_id}")
            print(f"Carrier: {label.ship_label_metadata.carrier_name}")
            
            # Download and open the shipping label
            self._download_and_open_label(label.file_data)
        else:
            print("⚠️  No labels generated")
        
        return response
    
    def _download_and_open_label(self, file_data) -> None:
        """
        Download and open the shipping label from the presigned URL.
        
        The file type is determined from the Content-Type header in the response:
        - x.application/zpl: ZPL format (for thermal printers)
        - application/pdf: PDF format
        - image/png: PNG format
        - text/plain: Plain text format
        
        In production, ZPL labels should be sent to a thermal printer.
        For this recipe, we download and open the label for demonstration.x
        
        """
        try:
            print(f"Downloading label...")
            
            # Download the label and get content type from response header
            with urllib.request.urlopen(file_data.url) as response:
                label_data = response.read()
                content_type = response.headers.get('Content-Type', 'application/octet-stream')
            
            # Determine file extension from content type
            if 'zpl' in content_type.lower():
                file_extension = '.zpl'
                label_format = 'ZPL'
            elif 'pdf' in content_type.lower():
                file_extension = '.pdf'
                label_format = 'PDF'
            elif 'png' in content_type.lower():
                file_extension = '.png'
                label_format = 'PNG'
            else:
                file_extension = '.txt'
                label_format = 'TEXT'
            
            print(f"Label Format: {label_format} (Content-Type: {content_type})")
            
            # Save to temporary file
            temp_file = tempfile.NamedTemporaryFile(
                delete=False,
                suffix=file_extension,
                prefix='shipping_label_'
            )
            temp_file.write(label_data)
            temp_file.close()
            
            print(f"✅ Label saved to: {temp_file.name}")
            
            # In production: Send ZPL files to thermal printer
            if label_format == 'ZPL':
                print(f"ℹ️  Production: Send ZPL file to thermal printer")
            
            # For this recipe: Open the downloaded file
            print(f"🖼️  Opening label...")
            webbrowser.open(f'file://{temp_file.name}')
            
        except Exception as e:
            print(f"❌ Error: {e}")

    # -------------------------------------------------------------------------
    # Main recipe entry point
    # -------------------------------------------------------------------------

    def start(self) -> None:
        """
        Main entry point showing the complete Easy Ship order processing flow.
        """
        print("=" * 70)
        print("Easy Ship Order Processing Recipe")
        print("=" * 70)
        
        # Step 1: Fetch open orders
        self.fetch_open_shipments()
        
        # Step 2: Acknowledge shipment
        self.acknowledge_shipment()
        
        # Step 3: Create packages
        self.create_packages()
        
        # Step 4: Retrieve shipping options
        self.retrieve_shipping_options()
        
        # Step 5: Generate shipping labels
        self.generate_shipping_labels()
        
        print("\n" + "=" * 70)
        print("✅ Easy Ship order processing completed successfully")
        print("=" * 70)


# -----------------------------------------------------------------------------
# Example usage for local/manual runs
# -----------------------------------------------------------------------------

if __name__ == "__main__":
    recipe = EasyShipOrderProcessingRecipe()
    recipe.start()
